src/WebJobs.Script/Binding/FunctionBinding.cs

```diff
@@ -133,5 +136,60 @@ internal static void AddStorageAccountAttribute(Collection<CustomAttributeBuilde
             var attribute = new CustomAttributeBuilder(typeof(StorageAccountAttribute).GetConstructor(constructorTypes), constructorArguments);
             attributes.Add(attribute);
         }
+
+        protected static ICollection<JToken> ReadAsCollection(Stream valueStream)
+        {
+            byte[] bytes;
+            using (MemoryStream ms = new MemoryStream())
+            {
+                valueStream.CopyTo(ms);
+                bytes = ms.ToArray();
+            }
+
+            JArray values = null;
+            string stringValue = Encoding.UTF8.GetString(bytes);
+
+            if (Utility.IsJson(stringValue))
+            {
+                JToken token = JToken.Parse(stringValue);
+                values = token as JArray;
+                if (token.Type != JTokenType.Array)
+                {
+                    values = new JArray();
+                    values.Add(token);
+                }
+            }
+            else
+            {
+                values = new JArray();
+                values.Add(stringValue);
+            }
+
+            return values;
+        }
+
+        protected async Task BindAsyncCollectorAsync<T>(Stream stream, IBinderEx binder, RuntimeBindingContext runtimeContext)
+        {
+            IAsyncCollector<T> collector = await binder.BindAsync<IAsyncCollector<T>>(runtimeContext);
+            ICollection<JToken> values = ReadAsCollection(stream);
+            foreach (var value in values)
+            {
+                object converted = null;
+                if (typeof(T) == typeof(string))
+                {
+                    converted = value.ToString();
+                }
+                else if (typeof(T) == typeof(JObject))
+                {
+                    converted = (JObject)value;
+                }
+                else
+                {
+                    throw new ArgumentException("Unsupported collection type.");
+                }
+
```

Comment:
Could this be simplified by doing the type check once, outside of the loop, just to throw if not a supported type and skip the conversion, directly casting the value to T?
I may be missing something, but if the expectation here is that in order to operate with bindings that expect strings and call BindAsyncCollectorAsync<string> (e.g Event Hub Biding), user code would need to return a string or an array of strings, that means we'd always be dealing with a JArray of strings when T is string, and it should be safe to cast perform that cast, so we don't need to call ToString.